File: src/VbNet.LanguageServer/Services/SemanticTokensService.cs
// SemanticTokensService - Provides semantic tokens via LSP
// Services Layer as defined in docs/architecture.md Section 5.4

using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Classification;
using Microsoft.CodeAnalysis.Text;
using Microsoft.Extensions.Logging;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Workspace;

namespace VbNet.LanguageServer.Services;

/// <summary>
/// Provides semantic tokens for VB.NET documents using Roslyn classifiers.
/// </summary>
public sealed class SemanticTokensService
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly ILogger<SemanticTokensService> _logger;

    private static readonly string[] TokenTypes = new[]
    {
        "namespace",
        "type",
        "class",
        "struct",
        "interface",
        "enum",
        "typeParameter",
        "function",
        "method",
        "property",
        "variable",
        "parameter",
        "field",
        "event",
        "keyword",
        "comment",
        "string",
        "number",
        "operator"
    };

    private static readonly string[] TokenModifiers = new[]
    {
        "declaration",
        "static",
        "readonly"
    };

    private static readonly Dictionary<string, TokenInfo> ClassificationMap = new(StringComparer.Ordinal)
    {
        [ClassificationTypeNames.NamespaceName] = new TokenInfo("namespace"),
        [ClassificationTypeNames.ClassName] = new TokenInfo("class"),
        [ClassificationTypeNames.StructName] = new TokenInfo("struct"),
        [ClassificationTypeNames.InterfaceName] = new TokenInfo("interface"),
        [ClassificationTypeNames.EnumName] = new TokenInfo("enum"),
        [ClassificationTypeNames.TypeParameterName] = new TokenInfo("typeParameter"),
        [ClassificationTypeNames.DelegateName] = new TokenInfo("type"),
        [ClassificationTypeNames.ModuleName] = new TokenInfo("type"),
        [ClassificationTypeNames.MethodName] = new TokenInfo("method"),
        [ClassificationTypeNames.ExtensionMethodName] = new TokenInfo("method"),
        [ClassificationTypeNames.PropertyName] = new TokenInfo("property"),
        [ClassificationTypeNames.FieldName] = new TokenInfo("field"),
        [ClassificationTypeNames.EventName] = new TokenInfo("event"),
        [ClassificationTypeNames.ParameterName] = new TokenInfo("parameter"),
        [ClassificationTypeNames.LocalName] = new TokenInfo("variable"),
        [ClassificationTypeNames.ConstantName] = new TokenInfo("variable", modifier: TokenModifier.Readonly),
        [ClassificationTypeNames.Keyword] = new TokenInfo("keyword"),
        [ClassificationTypeNames.ControlKeyword] = new TokenInfo("keyword"),
        [ClassificationTypeNames.Comment] = new TokenInfo("comment"),
        [ClassificationTypeNames.StringLiteral] = new TokenInfo("string"),
        [ClassificationTypeNames.VerbatimStringLiteral] = new TokenInfo("string"),
        [ClassificationTypeNames.NumericLiteral] = new TokenInfo("number"),
        [ClassificationTypeNames.Operator] = new TokenInfo("operator"),
        [ClassificationTypeNames.OperatorOverloaded] = new TokenInfo("operator")
    };

    public SemanticTokensService(
        WorkspaceManager workspaceManager,
        DocumentManager documentManager,
        ILogger<SemanticTokensService> logger)
    {
        _workspaceManager = workspaceManager ?? throw new ArgumentNullException(nameof(workspaceManager));
        _documentManager = documentManager ?? throw new ArgumentNullException(nameof(documentManager));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public static SemanticTokensOptions GetDefaultOptions()
    {
        return new SemanticTokensOptions
        {
            Legend = GetLegend(),
            Full = true,
            Range = true
        };
    }

    public static SemanticTokensLegend GetLegend()
    {
        return new SemanticTokensLegend
        {
            TokenTypes = TokenTypes,
            TokenModifiers = TokenModifiers
        };
    }

    public async Task<SemanticTokens> GetSemanticTokensAsync(SemanticTokensParams @params, CancellationToken cancellationToken)
    {
        if (@params?.TextDocument == null)
        {
            return new SemanticTokens();
        }

        var uri = @params.TextDocument.Uri;
        _logger.LogDebug("Semantic tokens requested for {Uri}", uri);

        var document = _documentManager.GetRoslynDocument(uri);
        if (document == null)
        {
            _logger.LogTrace("No Roslyn document found for: {Uri}", uri);
            return new SemanticTokens();
        }

        var sourceText = await document.GetTextAsync(cancellationToken);
        var span = new TextSpan(0, sourceText.Length);
        return await BuildTokensAsync(document, sourceText, span, cancellationToken);
    }

    public async Task<SemanticTokens> GetSemanticTokensRangeAsync(
        SemanticTokensRangeParams @params,
        CancellationToken cancellationToken)
    {
        if (@params?.TextDocument == null || @params.Range == null)
        {
            return new SemanticTokens();
        }

        var uri = @params.TextDocument.Uri;
        var document = _documentManager.GetRoslynDocument(uri);
        if (document == null)
        {
            _logger.LogTrace("No Roslyn document found for: {Uri}", uri);
            return new SemanticTokens();
        }

        var sourceText = await document.GetTextAsync(cancellationToken);
        var span = ToTextSpan(@params.Range, sourceText);
        return await BuildTokensAsync(document, sourceText, span, cancellationToken);
    }

    private static TextSpan ToTextSpan(Protocol.Range range, SourceText text)
    {
        var start = GetOffset(range.Start, text);
        var end = GetOffset(range.End, text);
        if (end < start)
        {
            (start, end) = (end, start);
        }

        return TextSpan.FromBounds(start, end);
    }

    private async Task<SemanticTokens> BuildTokensAsync(
        Document document,
        SourceText sourceText,
        TextSpan span,
        CancellationToken cancellationToken)
    {
        var classified = await Classifier.GetClassifiedSpansAsync(
            document,
            span,
            cancellationToken);

        if (!classified.Any())
        {
            return new SemanticTokens();
        }

        var tokens = new List<SemanticTokenData>();

        foreach (var item in classified)
        {
            if (!ClassificationMap.TryGetValue(item.ClassificationType, out var info))
            {
                continue;
            }

            if (!TryGetTokenTypeIndex(info.TokenType, out var tokenTypeIndex))
            {
                continue;
            }

            AppendTokenSegments(tokens, item.TextSpan, tokenTypeIndex, info.Modifier, sourceText);
        }

        tokens.Sort(SemanticTokenDataComparer.Instance);

        var data = EncodeTokens(tokens);
        return new SemanticTokens { Data = data };
    }

    private static void AppendTokenSegments(
        List<SemanticTokenData> tokens,
        TextSpan span,
        uint tokenType,
        TokenModifier modifier,
        SourceText sourceText)
    {
        var startLine = sourceText.Lines.GetLineFromPosition(span.Start);
        var endLine = sourceText.Lines.GetLineFromPosition(span.End);

        for (var line = startLine.LineNumber; line <= endLine.LineNumber; line++)
        {
            var textLine = sourceText.Lines[line];
            var lineStart = textLine.Start;
            var segmentStart = line == startLine.LineNumber ? span.Start - lineStart : 0;
            var segmentEnd = line == endLine.LineNumber ? span.End - lineStart : textLine.End - lineStart;
            var length = segmentEnd - segmentStart;
            if (length <= 0)
            {
                continue;
            }

            tokens.Add(new SemanticTokenData
            {
                Line = line,
                StartChar = segmentStart,
                Length = length,
                TokenType = tokenType,
                TokenModifiers = EncodeModifiers(modifier)
            });
        }
    }

    private static uint[] EncodeTokens(IReadOnlyList<SemanticTokenData> tokens)
    {
        var data = new List<uint>(tokens.Count * 5);
        var prevLine = 0;
        var prevStart = 0;

        foreach (var token in tokens)
        {
            var deltaLine = token.Line - prevLine;
            var deltaStart = deltaLine == 0 ? token.StartChar - prevStart : token.StartChar;

            data.Add((uint)deltaLine);
            data.Add((uint)deltaStart);
            data.Add((uint)token.Length);
            data.Add(token.TokenType);
            data.Add(token.TokenModifiers);

            prevLine = token.Line;
            prevStart = token.StartChar;
        }

        return data.ToArray();
    }

    private static uint EncodeModifiers(TokenModifier modifier)
    {
        return modifier switch
        {
            TokenModifier.Readonly => 1u << 2,
            _ => 0
        };
    }

    private static bool TryGetTokenTypeIndex(string tokenType, out uint index)
    {
        var idx = Array.IndexOf(TokenTypes, tokenType);
        if (idx < 0)
        {
            index = 0;
            return false;
        }

        index = (uint)idx;
        return true;
    }

    private static int GetOffset(Position position, SourceText text)
    {
        var line = Math.Min(position.Line, text.Lines.Count - 1);
        line = Math.Max(0, line);

        var textLine = text.Lines[line];
        var character = Math.Min(position.Character, textLine.End - textLine.Start);
        character = Math.Max(0, character);

        return textLine.Start + character;
    }

    private sealed class SemanticTokenDataComparer : IComparer<SemanticTokenData>
    {
        public static readonly SemanticTokenDataComparer Instance = new();

        public int Compare(SemanticTokenData x, SemanticTokenData y)
        {
            var line = x.Line.CompareTo(y.Line);
            if (line != 0)
            {
                return line;
            }

            return x.StartChar.CompareTo(y.StartChar);
        }
    }

    private readonly struct SemanticTokenData
    {
        public int Line { get; init; }
        public int StartChar { get; init; }
        public int Length { get; init; }
        public uint TokenType { get; init; }
        public uint TokenModifiers { get; init; }
    }

    private readonly struct TokenInfo
    {
        public TokenInfo(string tokenType, TokenModifier modifier = TokenModifier.None)
        {
            TokenType = tokenType;
            Modifier = modifier;
        }

        public string TokenType { get; }
        public TokenModifier Modifier { get; }
    }

    private enum TokenModifier
    {
        None,
        Readonly
    }
}

File: src/VbNet.LanguageServer.Vb/Services/SemanticTokensService.vb
' SemanticTokensService - Provides semantic tokens via LSP
' Services Layer as defined in docs/architecture.md Section 5.4

Imports System.Linq
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Classification
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.Extensions.Logging
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Workspace

Namespace Services

    ''' <summary>
    ''' Provides semantic tokens for VB.NET documents using Roslyn classifiers.
    ''' </summary>
    Public NotInheritable Class SemanticTokensService
        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _logger As ILogger(Of SemanticTokensService)

        Private Shared ReadOnly TokenTypes As String() = {
            "namespace",
            "type",
            "class",
            "struct",
            "interface",
            "enum",
            "typeParameter",
            "function",
            "method",
            "property",
            "variable",
            "parameter",
            "field",
            "event",
            "keyword",
            "comment",
            "string",
            "number",
            "operator"
        }

        Private Shared ReadOnly TokenModifiers As String() = {
            "declaration",
            "static",
            "readonly"
        }

        Private Shared ReadOnly ClassificationMap As Dictionary(Of String, TokenInfo) = New Dictionary(Of String, TokenInfo)(StringComparer.Ordinal) From {
            {ClassificationTypeNames.NamespaceName, New TokenInfo("namespace")},
            {ClassificationTypeNames.ClassName, New TokenInfo("class")},
            {ClassificationTypeNames.StructName, New TokenInfo("struct")},
            {ClassificationTypeNames.InterfaceName, New TokenInfo("interface")},
            {ClassificationTypeNames.EnumName, New TokenInfo("enum")},
            {ClassificationTypeNames.TypeParameterName, New TokenInfo("typeParameter")},
            {ClassificationTypeNames.DelegateName, New TokenInfo("type")},
            {ClassificationTypeNames.ModuleName, New TokenInfo("type")},
            {ClassificationTypeNames.MethodName, New TokenInfo("method")},
            {ClassificationTypeNames.ExtensionMethodName, New TokenInfo("method")},
            {ClassificationTypeNames.PropertyName, New TokenInfo("property")},
            {ClassificationTypeNames.FieldName, New TokenInfo("field")},
            {ClassificationTypeNames.EventName, New TokenInfo("event")},
            {ClassificationTypeNames.ParameterName, New TokenInfo("parameter")},
            {ClassificationTypeNames.LocalName, New TokenInfo("variable")},
            {ClassificationTypeNames.ConstantName, New TokenInfo("variable", TokenModifier.[ReadOnly])},
            {ClassificationTypeNames.Keyword, New TokenInfo("keyword")},
            {ClassificationTypeNames.ControlKeyword, New TokenInfo("keyword")},
            {ClassificationTypeNames.Comment, New TokenInfo("comment")},
            {ClassificationTypeNames.StringLiteral, New TokenInfo("string")},
            {ClassificationTypeNames.VerbatimStringLiteral, New TokenInfo("string")},
            {ClassificationTypeNames.NumericLiteral, New TokenInfo("number")},
            {ClassificationTypeNames.Operator, New TokenInfo("operator")},
            {ClassificationTypeNames.OperatorOverloaded, New TokenInfo("operator")}
        }

        Public Sub New(workspaceManager As WorkspaceManager, documentManager As DocumentManager, logger As ILogger(Of SemanticTokensService))
            If workspaceManager Is Nothing Then
                Throw New ArgumentNullException(NameOf(workspaceManager))
            End If
            If documentManager Is Nothing Then
                Throw New ArgumentNullException(NameOf(documentManager))
            End If
            If logger Is Nothing Then
                Throw New ArgumentNullException(NameOf(logger))
            End If

            _workspaceManager = workspaceManager
            _documentManager = documentManager
            _logger = logger
        End Sub

        Public Shared Function GetDefaultOptions() As SemanticTokensOptions
            Return New SemanticTokensOptions With {
                .Legend = GetLegend(),
                .Full = True,
                .Range = True
            }
        End Function

        Public Shared Function GetLegend() As SemanticTokensLegend
            Return New SemanticTokensLegend With {
                .TokenTypes = TokenTypes,
                .TokenModifiers = TokenModifiers
            }
        End Function

        Public Async Function GetSemanticTokensAsync(parameters As SemanticTokensParams, cancellationToken As CancellationToken) As Task(Of SemanticTokens)
            If parameters Is Nothing OrElse parameters.TextDocument Is Nothing Then
                Return New SemanticTokens()
            End If

            Dim uri = parameters.TextDocument.Uri
            _logger.LogDebug("Semantic tokens requested for {Uri}", uri)

            Dim document = _documentManager.GetRoslynDocument(uri)
            If document Is Nothing Then
                _logger.LogTrace("No Roslyn document found for: {Uri}", uri)
                Return New SemanticTokens()
            End If

            Dim sourceText = Await document.GetTextAsync(cancellationToken).ConfigureAwait(False)
            Dim span As New TextSpan(0, sourceText.Length)
            Return Await BuildTokensAsync(document, sourceText, span, cancellationToken).ConfigureAwait(False)
        End Function

        Public Async Function GetSemanticTokensRangeAsync(parameters As SemanticTokensRangeParams, cancellationToken As CancellationToken) As Task(Of SemanticTokens)
            If parameters Is Nothing OrElse parameters.TextDocument Is Nothing OrElse parameters.Range Is Nothing Then
                Return New SemanticTokens()
            End If

            Dim uri = parameters.TextDocument.Uri
            Dim document = _documentManager.GetRoslynDocument(uri)
            If document Is Nothing Then
                _logger.LogTrace("No Roslyn document found for: {Uri}", uri)
                Return New SemanticTokens()
            End If

            Dim sourceText = Await document.GetTextAsync(cancellationToken).ConfigureAwait(False)
            Dim span = ToTextSpan(parameters.Range, sourceText)
            Return Await BuildTokensAsync(document, sourceText, span, cancellationToken).ConfigureAwait(False)
        End Function

        Private Shared Function ToTextSpan(range As Protocol.Range, text As SourceText) As TextSpan
            Dim start = GetOffset(range.Start, text)
            Dim [end] = GetOffset(range.End, text)
            If [end] < start Then
                Dim tmp = start
                start = [end]
                [end] = tmp
            End If

            Return TextSpan.FromBounds(start, [end])
        End Function

        Private Async Function BuildTokensAsync(document As Document, sourceText As SourceText, span As TextSpan, cancellationToken As CancellationToken) As Task(Of SemanticTokens)
            Dim classified = Await Classifier.GetClassifiedSpansAsync(document, span, cancellationToken).ConfigureAwait(False)

            If Not classified.Any() Then
                Return New SemanticTokens()
            End If

            Dim tokens As New List(Of SemanticTokenData)()

            For Each item In classified
                Dim info As TokenInfo = Nothing
                If Not ClassificationMap.TryGetValue(item.ClassificationType, info) Then
                    Continue For
                End If

                Dim tokenTypeIndex As UInteger
                If Not TryGetTokenTypeIndex(info.TokenType, tokenTypeIndex) Then
                    Continue For
                End If

                AppendTokenSegments(tokens, item.TextSpan, tokenTypeIndex, info.Modifier, sourceText)
            Next

            tokens.Sort(SemanticTokenDataComparer.Instance)

            Dim data = EncodeTokens(tokens)
            Return New SemanticTokens With {.Data = data}
        End Function

        Private Shared Sub AppendTokenSegments(tokens As List(Of SemanticTokenData), span As TextSpan, tokenType As UInteger, modifier As TokenModifier, sourceText As SourceText)
            Dim startLine = sourceText.Lines.GetLineFromPosition(span.Start)
            Dim endLine = sourceText.Lines.GetLineFromPosition(span.[End])

            For line = startLine.LineNumber To endLine.LineNumber
                Dim textLine = sourceText.Lines(line)
                Dim lineStart = textLine.Start
                Dim segmentStart = If(line = startLine.LineNumber, span.Start - lineStart, 0)
                Dim segmentEnd = If(line = endLine.LineNumber, span.[End] - lineStart, textLine.End - lineStart)
                Dim length = segmentEnd - segmentStart
                If length <= 0 Then
                    Continue For
                End If

                tokens.Add(New SemanticTokenData With {
                    .Line = line,
                    .StartChar = segmentStart,
                    .Length = length,
                    .TokenType = tokenType,
                    .TokenModifiers = EncodeModifiers(modifier)
                })
            Next
        End Sub

        Private Shared Function EncodeTokens(tokens As IReadOnlyList(Of SemanticTokenData)) As UInteger()
            Dim data As New List(Of UInteger)(tokens.Count * 5)
            Dim prevLine = 0
            Dim prevStart = 0

            For Each token In tokens
                Dim deltaLine = token.Line - prevLine
                Dim deltaStart = If(deltaLine = 0, token.StartChar - prevStart, token.StartChar)

                data.Add(CUInt(deltaLine))
                data.Add(CUInt(deltaStart))
                data.Add(CUInt(token.Length))
                data.Add(token.TokenType)
                data.Add(token.TokenModifiers)

                prevLine = token.Line
                prevStart = token.StartChar
            Next

            Return data.ToArray()
        End Function

        Private Shared Function EncodeModifiers(modifier As TokenModifier) As UInteger
            Select Case modifier
                Case TokenModifier.[ReadOnly]
                    Return CUInt(1) << 2
                Case Else
                    Return 0
            End Select
        End Function

        Private Shared Function TryGetTokenTypeIndex(tokenType As String, ByRef index As UInteger) As Boolean
            Dim idx = Array.IndexOf(TokenTypes, tokenType)
            If idx < 0 Then
                index = 0
                Return False
            End If

            index = CUInt(idx)
            Return True
        End Function

        Private Shared Function GetOffset(position As Position, text As SourceText) As Integer
            Dim line = Math.Min(position.Line, text.Lines.Count - 1)
            line = Math.Max(0, line)

            Dim textLine = text.Lines(line)
            Dim character = Math.Min(position.Character, textLine.End - textLine.Start)
            character = Math.Max(0, character)

            Return textLine.Start + character
        End Function

        Private NotInheritable Class SemanticTokenDataComparer
            Implements IComparer(Of SemanticTokenData)

            Public Shared ReadOnly Instance As SemanticTokenDataComparer = New SemanticTokenDataComparer()

            Public Function Compare(x As SemanticTokenData, y As SemanticTokenData) As Integer Implements IComparer(Of SemanticTokenData).Compare
                Dim line = x.Line.CompareTo(y.Line)
                If line <> 0 Then
                    Return line
                End If

                Return x.StartChar.CompareTo(y.StartChar)
            End Function
        End Class

        Private Structure SemanticTokenData
            Public Property Line As Integer
            Public Property StartChar As Integer
            Public Property Length As Integer
            Public Property TokenType As UInteger
            Public Property TokenModifiers As UInteger
        End Structure

        Private Structure TokenInfo
            Public Sub New(tokenType As String, Optional modifier As TokenModifier = TokenModifier.None)
                Me.TokenType = tokenType
                Me.Modifier = modifier
            End Sub

            Public ReadOnly Property TokenType As String
            Public ReadOnly Property Modifier As TokenModifier
        End Structure

        Private Enum TokenModifier
            None
            [ReadOnly]
        End Enum
    End Class

End Namespace
